Drop BufferSource::data() in favor of span()

cdumez opened the pull request:
#### c98317135048a7921f493b04679f707e4ac4f197
<pre>
Drop BufferSource::data() in favor of span()
<a href="https://bugs.webkit.org/show_bug.cgi?id=275796">https://bugs.webkit.org/show_bug.cgi?id=275796</a>

Reviewed by Darin Adler.

* Source/JavaScriptCore/runtime/ArrayBuffer.h:
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::appendFileContentsToFileHandle):
(WTF::FileSystemImpl::readOrMakeSalt):
(WTF::FileSystemImpl::readEntireFile):
* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::readFromFile):
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::readFromFile):
* Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.cpp:
(WebCore::FileSystemSyncAccessHandle::read):
* Source/WebCore/Modules/webcodecs/WebCodecsAudioData.cpp:
(WebCore::WebCodecsAudioData::copyTo):
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
(WebCore::WebCodecsAudioDecoder::decode):
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedAudioChunk.cpp:
(WebCore::WebCodecsEncodedAudioChunk::copyTo):
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedAudioChunk.h:
(WebCore::WebCodecsEncodedAudioChunk::span const):
(WebCore::WebCodecsEncodedAudioChunk::data const): Deleted.
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.cpp:
(WebCore::WebCodecsEncodedVideoChunk::copyTo):
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.h:
(WebCore::WebCodecsEncodedVideoChunk::span const):
(WebCore::WebCodecsEncodedVideoChunk::data const): Deleted.
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
(WebCore::WebCodecsVideoDecoder::decode):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
(WebCore::WebCodecsVideoFrame::create):
(WebCore::WebCodecsVideoFrame::copyTo):
* Source/WebCore/bindings/js/BufferSource.h:
(WebCore::BufferSource::length const):
(WebCore::BufferSource::span const):
(WebCore::BufferSource::mutableSpan):
(WebCore::toNSData):
(WebCore::BufferSource::data const): Deleted.
(WebCore::BufferSource::mutableData const): Deleted.
* Source/WebCore/platform/FileHandle.cpp:
(WebCore::FileHandle::read):
* Source/WebCore/platform/FileStream.cpp:
(WebCore::FileStream::read):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
(WebCore::unmapFunction):
* Source/WebCore/platform/network/curl/CurlFormDataStream.cpp:
(WebCore::CurlFormDataStream::readFromFile):
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelCurl.cpp:
(WebKit::NetworkCache::IOChannel::read):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
(WebKit::LocalAuthenticator::deleteDuplicateCredential const):
* Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GstElementHarness.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: <a href="https://commits.webkit.org/280309@main">https://commits.webkit.org/280309@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4d39ccfe535094bfc91e2dcf97f7797d5f1184cb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56241 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35567 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8713 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/59847 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6677 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58367 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43189 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6871 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45535 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4643 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58270 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33452 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48520 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26418 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30232 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5851 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/5681 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/49319 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6123 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61530 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/55478 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/149 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6249 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52814 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/149 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48587 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/52693 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/132 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/77238 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8345 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31394 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12800 "Found 1 new JSC stress test failure: stress/try-get-value-without-gc.js.default (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32480 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33563 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32227 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/bindings/js/BufferSource.h:
```diff
@@ -52,29 +52,26 @@ class BufferSource {
     {
-        return std::visit([](auto& buffer) -> const uint8_t* {
-            return buffer ? static_cast<const uint8_t*>(buffer->data()) : nullptr;
+        return std::visit([](auto& buffer) -> size_t {
```
Does this compile without the explicit `size_t`?

darinadler commented on Source/WebCore/bindings/js/BufferSource.h:
```diff
@@ -92,7 +89,8 @@ inline BufferSource toBufferSource(NSData *data)
 {
-    return adoptNS([[NSData alloc] initWithBytes:data.data() length:data.length()]);
+    auto span = data.span();
+    return adoptNS([[NSData alloc] initWithBytes:span.data() length:span.size()]);
```
Don't we have a function that does this already that we can call?

cdumez replied:
I added `toNSData()` in SpanCocoa.h.

darinadler commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -1465,12 +1466,12 @@ HTTPServer server([parentFrame = String(parentFrame), subFrame = String(subFrame
 {
-    uint8_t identifier[] = { 0x01, 0x02, 0x03, 0x04 };
-    NSData *nsIdentifier = [NSData dataWithBytes:identifier length:sizeof(identifier)];
+    std::array<uint8_t, 4> identifier { 0x01, 0x02, 0x03, 0x04 };
```
Doesn't it deduce the `4` if we omit it?

cdumez replied:
We can’t simply omit the 4 though, we’d need to omit uint8_t too and then it would deduce the wrong type. There is std::to_array iirc the name but the Sony port doesn’t support it yet.

cdumez replied:
This is Cocoa code so I'll use `std::to_array`

darinadler commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -1465,12 +1466,12 @@ HTTPServer server([parentFrame = String(parentFrame), subFrame = String(subFrame
-    uint8_t identifier[] = { 0x01, 0x02, 0x03, 0x04 };
-    NSData *nsIdentifier = [NSData dataWithBytes:identifier length:sizeof(identifier)];
+    std::array<uint8_t, 4> identifier { 0x01, 0x02, 0x03, 0x04 };
+    RetainPtr nsIdentifier = adoptNS([[NSData alloc] initWithBytes:identifier.data() length:identifier.size()]);
```
Don't we have a helper function to make an `NSData` by copying from a span?

cdumez replied:
I will check. I am familiar with the one creating a span from NSData, not the other way around.

cdumez committed:
Drop BufferSource::data() in favor of span()
https://bugs.webkit.org/show_bug.cgi?id=275796

Reviewed by Darin Adler.

* Source/JavaScriptCore/runtime/ArrayBuffer.h:
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::appendFileContentsToFileHandle):
(WTF::FileSystemImpl::readOrMakeSalt):
(WTF::FileSystemImpl::readEntireFile):
* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::readFromFile):
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::readFromFile):
* Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.cpp:
(WebCore::FileSystemSyncAccessHandle::read):
* Source/WebCore/Modules/webcodecs/WebCodecsAudioData.cpp:
(WebCore::WebCodecsAudioData::copyTo):
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
(WebCore::WebCodecsAudioDecoder::decode):
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedAudioChunk.cpp:
(WebCore::WebCodecsEncodedAudioChunk::copyTo):
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedAudioChunk.h:
(WebCore::WebCodecsEncodedAudioChunk::span const):
(WebCore::WebCodecsEncodedAudioChunk::data const): Deleted.
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.cpp:
(WebCore::WebCodecsEncodedVideoChunk::copyTo):
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.h:
(WebCore::WebCodecsEncodedVideoChunk::span const):
(WebCore::WebCodecsEncodedVideoChunk::data const): Deleted.
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
(WebCore::WebCodecsVideoDecoder::decode):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
(WebCore::WebCodecsVideoFrame::create):
(WebCore::WebCodecsVideoFrame::copyTo):
* Source/WebCore/bindings/js/BufferSource.h:
(WebCore::BufferSource::length const):
(WebCore::BufferSource::span const):
(WebCore::BufferSource::mutableSpan):
(WebCore::toNSData):
(WebCore::BufferSource::data const): Deleted.
(WebCore::BufferSource::mutableData const): Deleted.
* Source/WebCore/platform/FileHandle.cpp:
(WebCore::FileHandle::read):
* Source/WebCore/platform/FileStream.cpp:
(WebCore::FileStream::read):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
(WebCore::unmapFunction):
* Source/WebCore/platform/network/curl/CurlFormDataStream.cpp:
(WebCore::CurlFormDataStream::readFromFile):
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelCurl.cpp:
(WebKit::NetworkCache::IOChannel::read):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
(WebKit::LocalAuthenticator::deleteDuplicateCredential const):
* Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GstElementHarness.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: https://commits.webkit.org/280309@main
